src/app/CASESessionManager.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/CASESessionManager.h>
+
+namespace chip {
+
+CHIP_ERROR CASESessionManager::FindOrEstablishSession(NodeId deviceId, Transport::PeerAddress address,
+                                                      Callback::Callback<OnDeviceConnected> * onConnection,
+                                                      Callback::Callback<OnDeviceConnectionFailure> * onFailure)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+    PeerId peerId                    = mInitParams.sessionInitParams.fabricInfo->GetPeerIdForNode(deviceId);
+    OperationalDeviceProxy * session = FindSession(deviceId);
+    if (session == nullptr)
+    {
+        session = mActiveSessions.CreateObject(mInitParams.sessionInitParams, peerId);
+        if (session == nullptr)
+        {
+            onFailure->mCall(onFailure->mContext, deviceId, CHIP_ERROR_NO_MEMORY);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+    }
+
+    bool validAddress = (address != Transport::PeerAddress::UDP(Inet::IPAddress::Any));
+
+    if (!validAddress && mInitParams.dnsCache != nullptr)
+    {
+        Dnssd::ResolvedNodeData nodeData;
+        if (mInitParams.dnsCache->Lookup(peerId, nodeData) == CHIP_NO_ERROR)
+        {
+            address      = ToPeerAddress(nodeData);
+            validAddress = true;
+        }
+    }
+
+    if (validAddress)
+    {
+        uint32_t idleInterval;
+        uint32_t activeInterval;
+        session->GetMRPIntervals(idleInterval, activeInterval);
+        CHIP_ERROR err = session->UpdateDeviceData(address, idleInterval, activeInterval);
+        if (err != CHIP_NO_ERROR)
+        {
+            ReleaseSession(session);
```

Comment:
could you use some RAII and have a 'armrelease' or similar? this pattern of if(error) release and return seems to be reused and seems error prone if the size of this method increases.

Reply:
I reworked the code to remove the extra checks and returns.